Source/WebCore/platform/ios/ScreenOrientationProviderIOS.mm

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+
+#if PLATFORM(IOS_FAMILY)
+#import "ScreenOrientationProvider.h"
+
+#import <pal/ios/UIKitSoftLink.h>
+
+static WebCore::ScreenOrientationType currentOrientationFromCurrentUIDevice()
+{
+    switch ([[PAL::getUIDeviceClass() currentDevice] orientation]) {
+    case UIDeviceOrientationUnknown:
+    case UIDeviceOrientationFaceUp:
+    case UIDeviceOrientationFaceDown:
+    case UIDeviceOrientationPortrait:
+        break;
+    case UIDeviceOrientationPortraitUpsideDown:
+        return WebCore::ScreenOrientationType::PortraitSecondary;
+    case UIDeviceOrientationLandscapeLeft:
+        return WebCore::ScreenOrientationType::LandscapePrimary;
+    case UIDeviceOrientationLandscapeRight:
+        return WebCore::ScreenOrientationType::LandscapeSecondary;
+    }
+    return WebCore::ScreenOrientationType::PortraitPrimary;
+}
+
+@interface WebScreenOrientationObserver : NSObject
+@property (nonatomic) WebCore::ScreenOrientationProvider* provider;
+@end
+
+@implementation WebScreenOrientationObserver {
+}
+
+- (WebScreenOrientationObserver *)initWithProvider:(WebCore::ScreenOrientationProvider&)provider
+{
+    self = [super init];
+    if (!self)
+        return nil;
+
+    _provider = &provider;
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_screenOrientationDidChange) name:PAL::get_UIKit_UIDeviceOrientationDidChangeNotification() object:nil];
```

Comment:
Hm...I wonder if, instead of getting the device orientation from the notification, we should be getting it from the embedding window's `UIWindowScene` instead (`-[UIWindowScene interfaceOrientation]`). That behavior _might_ be more sensible in the case of external displays on iPad, for instance, if the web view is on the external display and you rotate the device.

I'm not totally sure we'd want to fire this notification in that case.